src/scale_impls/decode.rs

```diff
@@ -64,261 +28,209 @@ pub fn decode_value_as_type<Id: Into<TypeId>>(
+		type_id: scale_decode::visitor::TypeId,
+	) -> Result<Self::Value, Self::Error> {
+		to_unnamed_composite!(value, type_id)
+	}
+	fn visit_array(
+		self,
+		value: &mut scale_decode::visitor::Array,
+		type_id: scale_decode::visitor::TypeId,
+	) -> Result<Self::Value, Self::Error> {
+		to_unnamed_composite!(value, type_id)
+	}
+	fn visit_bitsequence(
+		self,
+		value: &mut scale_decode::visitor::BitSequence,
+		type_id: scale_decode::visitor::TypeId,
+	) -> Result<Self::Value, Self::Error> {
+		Ok(Value {
+			value: ValueDef::BitSequence(value.decode_bitsequence()?.to_u8_lsb0()),
+			context: type_id.into(),
+		})
+	}
+	fn visit_str(
+		self,
+		value: scale_decode::visitor::Str,
+		type_id: scale_decode::visitor::TypeId,
+	) -> Result<Self::Value, Self::Error> {
+		Ok(Value::string(value.as_str()?).map_context(|_| type_id.into()))
+	}
+	fn visit_variant(
+		self,
+		value: &mut scale_decode::visitor::Variant,
+		type_id: scale_decode::visitor::TypeId,
+	) -> Result<Self::Value, Self::Error> {
+		let values = visit_composite(value.fields())?;
+		Ok(Value {
+			value: ValueDef::Variant(Variant { name: value.name().to_owned(), values }),
+			context: type_id.into(),
+		})
+	}
+	fn visit_composite(
+		self,
+		value: &mut scale_decode::visitor::Composite,
+		type_id: scale_decode::visitor::TypeId,
+	) -> Result<Self::Value, Self::Error> {
+		Ok(Value { value: ValueDef::Composite(visit_composite(value)?), context: type_id.into() })
 	}
 }
 
-fn decode_sequence_value(
-	data: &mut &[u8],
-	ty: &TypeDefSequence<PortableForm>,
-	types: &PortableRegistry,
-) -> Result<Composite<TypeId>, DecodeError> {
-	// We assume that the sequence is preceeded by a compact encoded length, so that
-	// we know how many values to try pulling out of the data.
-	let len = Compact::<u64>::decode(data)?;
-	let values: Vec<_> = (0..len.0)
-		.map(|_| decode_value_as_type(data, ty.type_param(), types))
-		.collect::<Result<_, _>>()?;
-
-	Ok(Composite::Unnamed(values))
-}
-
-fn decode_array_value(
-	data: &mut &[u8],
-	ty: &TypeDefArray<PortableForm>,
-	types: &PortableRegistry,
-) -> Result<Composite<TypeId>, DecodeError> {
-	// The length is known based on the type we want to decode into, so we pull out the number of items according
-	// to that, and don't need a length to exist in the SCALE encoded bytes
-	let values: Vec<_> = (0..ty.len())
-		.map(|_| decode_value_as_type(data, ty.type_param(), types))
-		.collect::<Result<_, _>>()?;
-
-	Ok(Composite::Unnamed(values))
-}
-
-fn decode_tuple_value(
-	data: &mut &[u8],
-	ty: &TypeDefTuple<PortableForm>,
-	types: &PortableRegistry,
+/// Extract a named/unnamed Composite type out of scale_decode's Composite.
+fn visit_composite(
+	value: &mut scale_decode::visitor::Composite,
 ) -> Result<Composite<TypeId>, DecodeError> {
-	let values: Vec<_> = ty
-		.fields()
-		.iter()
-		.map(|f| decode_value_as_type(data, f, types))
-		.collect::<Result<_, _>>()?;
+	let len = value.fields().len();
```

Comment:
these are a bit redundant i.e, if `len > 0` then "named" should be true as well

in addition as both branches still creates new `Vec` I would prefer to skip the `if/else` completely

Reply:
Ah two things;
- Fields all come with a `.name()` which is `Option<String>`. We assume that if the first field isn't named (ie is `None`) then other fields will be unnamed too (it's either eg a named struct or a tuple). So if no fields exist we just output an empty `Composite::Unnamed` (but either is sortof fine). So it's more than just checking whether `len > 0`.
- `let mut vals = Vec::with_capacity(len);` looks identical on either branch but it's actually a different type in each one, so it has to be duplicated :)

Reply:
ah ok, I didn't read it carefully enough then cool